app/src/main/java/com/framgia/chat_03/screen/signup/SignUpActivity.java

```diff
@@ -31,4 +59,156 @@ protected void onStop() {
+        Toast.makeText(this, R.string.signup_confirm_password_error, Toast.LENGTH_SHORT).show();
+    }
+
+    @Override
+    public void onCreateNewAccountFail() {
+        Toast.makeText(this, R.string.signup_create_error, Toast.LENGTH_SHORT).show();
+    }
+
+    @Override
+    public void onWeakPassword() {
+        Toast.makeText(this, R.string.signup_weak_password, Toast.LENGTH_SHORT).show();
+    }
+
+    @Override
+    public void saveUser() {
+        String email = mEditTextEmail.getText().toString().trim();
+        String password = mEditTextPassword.getText().toString().trim();
+        String fullName = mEditTextFullName.getText().toString().trim();
+        User user = new User(email, fullName, password, null);
+        mPresenter.saveUser(user);
+    }
+
+    @Override
+    public void startHomeScreen() {
+
+    }
+
+    @Override
+    public boolean isCheckSelfPermission(String permission) {
+        return ContextCompat.checkSelfPermission(this, permission)
+                != PackageManager.PERMISSION_GRANTED;
+    }
+
+    @Override
+    public void chooseImageFromGallery() {
+        Intent intent = new Intent(Intent.ACTION_PICK);
+        intent.setType(TYPE_INTENT_PICK_IMAGE);
+        startActivityForResult(intent, PICK_IMAGE_FROM_GALLERY);
+    }
+
+    @Override
+    public void onPermissionError() {
+        Toast.makeText(this, getString(R.string.signup_permission_error),
+                Toast.LENGTH_SHORT).show();
+    }
+
+    @Override
+    public void requestPermission(String permission, int requestCode) {
+        ActivityCompat.requestPermissions(this, new String[]{permission},
+                requestCode);
+    }
+
+    @Override
+    public void setImageDialogFromGallery(Uri pathImage) {
+        Glide.with(this)
+                .load(pathImage)
+                .apply(RequestOptions.circleCropTransform())
+                .into(mImageViewAvatar);
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+                                           @NonNull int[] grantResults) {
+        mPresenter.onRequestPermissionsResult(requestCode, grantResults);
+    }
+
+    private void initViews() {
+        mPresenter.setView(this);
+        mDialogLogin = new ProgressDialog(this);
+        mEditTextFullName = findViewById(R.id.text_full_name);
+        mEditTextEmail = findViewById(R.id.text_email);
+        mEditTextPassword = findViewById(R.id.text_password);
+        mEditTextConfirmPassword = findViewById(R.id.text_re_password);
+        mImageViewAvatar = findViewById(R.id.image_avatar);
+        initComponent();
+    }
+
+    private void initComponent() {
+        findViewById(R.id.image_back).setOnClickListener(this);
+        findViewById(R.id.button_sign_up).setOnClickListener(this);
+        mImageViewAvatar.setOnClickListener(this);
+    }
+
+    private void initData() {
+        mPresenter = new SignUpPresenter(
```

Comment:
em nên tách các biến ra cho gọn
```suggestion
        mPresenter = new SignUpPresenter(authenticationRepository, userRepository)
```